framework/src/constraints/AutomaticMortarGeneration.C

```diff
@@ -1190,20 +1184,28 @@ AutomaticMortarGeneration::buildCouplingInformation()
     coupling_info[secondary_elem->processor_id()].emplace_back(
         secondary_elem->id(), secondary_elem->interior_parent()->id());
     if (secondary_elem->processor_id() != _mesh.processor_id())
-      // We want to keep information for nonlocal secondary element point neighbors for mortar nodal
-      // aux kernels
+      // We want to keep information for nonlocal lower-dimensional secondary element point
+      // neighbors for mortar nodal aux kernels
       _mortar_interface_coupling[secondary_elem->id()].insert(
           secondary_elem->interior_parent()->id());
 
     // LowerPrimary
     coupling_info[secondary_elem->processor_id()].emplace_back(
         secondary_elem->id(), primary_elem->interior_parent()->id());
     if (secondary_elem->processor_id() != _mesh.processor_id())
-      // We want to keep information for nonlocal secondary element point neighbors for mortar nodal
-      // aux kernels
+      // We want to keep information for nonlocal lower-dimensional secondary element point
+      // neighbors for mortar nodal aux kernels
       _mortar_interface_coupling[secondary_elem->id()].insert(
           primary_elem->interior_parent()->id());
 
+    // Lower-LowerDimensionalPrimary
+    coupling_info[secondary_elem->processor_id()].emplace_back(secondary_elem->id(),
+                                                               primary_elem->id());
+    if (secondary_elem->processor_id() != _mesh.processor_id())
+      // We want to keep information for nonlocal lower-dimensional secondary element point
+      // neighbors for mortar nodal aux kernels
+      _mortar_interface_coupling[secondary_elem->id()].insert(primary_elem->id());
```

Comment:
it wasn't just for mortar nodal aux kernels. We were missing the lower dimensional primary element ghosting in 3D for mortar constraints. We just never noticed it before because we never deleted remote elements. This use case was covered in the heat conduction module, so we do have coverage within the MOOSE repo. However, if you're asking whether we have a framework level test for that yet ... I think the answer is no

Reply:
> We were missing the lower dimensional primary element ghosting in 3D for mortar constraints

So the lower-d primary mesh suddenly stops (no ghosting) for parallel 3D applications? I guess we are not hitting this case from the contact module either.

Reply:
No. For some reason, which I should figure out, even if a user asks for `--distributed-mesh` with mortar they are not getting it in most cases

Reply:
This is the code in `AugmentSparsityOnInterface::operator()` which is responsible for most distributed mesh simulations not being truly distributed with respect to the elements contributing to mortar
```c++
  if ((!amg || _use_displaced_mesh) && !_is_coupling_functor)
  {
    for (const Elem * const elem : _mesh->active_element_ptr_range())
    {
      if (generating_mesh)
      {
        // We are still generating the mesh, so it's possible we don't even have the right boundary
        // ids created yet! So we actually ghost all boundary elements and all lower dimensional
        // elements who have parents on a boundary
```
`generating_mesh` is `true` while mesh generators are running, which is when `prepare_for_use/delete_remote_elements` gets called on the libMesh mesh(es). The addition of this new `GhostLowerDElems` RM means that in `SetupMeshCompleteAction` we now enter the conditional at the bottom here whenever there are `SideUserObjects` in the simulation (because this new RM is a late geometric ghosting RM):
```c++
  else if (_current_task == "delete_remote_elements_after_late_geometric_ghosting")
  {
    TIME_SECTION("deleteRemoteElems", 2, "Deleting Remote Elements");

    if (_displaced_mesh &&
        (_mesh->needsRemoteElemDeletion() != _displaced_mesh->needsRemoteElemDeletion()))
      mooseError("Our reference and displaced meshes are not in sync with respect to whether we "
                 "should delete remote elements.");

    // We currently only trigger the needsRemoteDeletion flag if somebody has requested a late
    // geometric ghosting functor and/or we have a displaced mesh
    if (_mesh->needsRemoteElemDeletion())
```

Reply:
I see. I'd figure the solution isn't as simple as delaying the call to  `AugmentSparsityOnInterface::operator()` even for cases that do not require `SetupMeshCompleteAction`.

Reply:
the solution to what? Ensuring that we actually delete remote elements for all distributed mesh simulations? Or something else?

Reply:
The solution `the mortar mesh becoming truly distributed` when not using `SideUserObjects`.

Reply:
I will make the geometric ghosting functor in `MortarConsumerInterface::validParams` a late one ... this should ensure that we always get a distributed mortar interface for undisplaced meshes. However for displaced meshes we will still ghost the entire mortar interface since we do not know how the mesh will move and consequently what projections we will have. The alternative to ghosting the entire mortar interface for displaced meshes would be to do communication of elements at each residual/Jacobian evaluation

Reply:
I'm not sure if I want to take that on right now even if it's a good idea (and I'm not sure whether the memory gains of having more distribution would be worth the communcation cpu costs)

Reply:
oh dear ... now it looks like there are issues with the VCP :frowning: